File: Divide and Conquer/Quicksort/Code.java
import org.algorithm_visualizer.*;
import java.util.Arrays;

class Main {
    //visualizer{
    private static ChartTracer chartTracer = new ChartTracer();
    private static LogTracer logTracer = new LogTracer("Console");
    private static Array1DTracer tracer = new Array1DTracer();
    //}
    private static Integer[] array = (Integer[]) new Randomize.Array1D(15, new Randomize.Integer(1, 20)).create();

    public static void main(String[] args) {
        //visualizer{
        tracer.set(array);
        tracer.chart(chartTracer);
        Layout.setRoot(new VerticalLayout(new Commander[]{chartTracer, tracer, logTracer}));
        logTracer.printf("original array = %s\n", Arrays.toString(array));

        Tracer.delay();
        //}
        quickSort(array, 0, array.length - 1);
        //visualizer{
        logTracer.printf("sorted array = %s\n", Arrays.toString(array));
        //}
    }

    public static void quickSort(Integer[] arr, int left, int right) {
        int l, r, s;
        while (right > left) {
            l = left;
            r = right;
            s = arr[left];
            while (l < r) {
                //visualizer{
                tracer.select(left);
                tracer.select(right);
                Tracer.delay();
                //}
                while (arr[r] > s) {
                    //visualizer{
                    tracer.select(r);
                    Tracer.delay();
                    tracer.deselect(r);
                    //}
                    r--;
                }
                arr[l] = arr[r];
                //visualizer{
                tracer.patch(l, arr[r]);
                Tracer.delay();
                tracer.depatch(l);
                //}
                while (s >= arr[l] && l < r) {
                    //visualizer{
                    tracer.select(l);
                    Tracer.delay();
                    tracer.deselect(l);
                    //}
                    l++;
                }
                arr[r] = arr[l];
                //visualizer{
                tracer.patch(r, arr[l]);
                Tracer.delay();
                tracer.depatch(r);
                tracer.deselect(left);
                tracer.deselect(right);
                //}
            }
            arr[l] = s;
            //visualizer{
            tracer.patch(l, s);
            Tracer.delay();
            tracer.depatch(l);
            //}
            quickSort(arr, left, l - 1);
            left = l + 1;
        }
    }
}

File: Divide and Conquer/Quicksort/code.cpp
#include "algorithm-visualizer.h"
#include <iostream>

//visualizer{
const int				ARRAYSIZE = 20;
ChartTracer				chart("ChartTracer");
Array1DTracer			tracer("Array1DTracer");
LogTracer				logger("LogTracer");
Randomize::Array1D<int> array1D(ARRAYSIZE);
int						D[ARRAYSIZE];

void InitQuickSort()
{
	Layout::setRoot(VerticalLayout({ chart,tracer,logger }));
	array1D.fill(D);
	tracer.set(D);
	tracer.chart(chart);
	Tracer::delay();
}
//}

void partition(int D[ARRAYSIZE], int low, int high)
{
	int i;
	int j;
	int s;
	while (high > low) {
		i = low;
		j = high;
		s = D[low];
		while (i < j) {
			//visualizer{
			tracer.select(high);
			tracer.select(low);
			Tracer::delay();
			//}
			while (D[j] > s) {
				//visualizer{
				tracer.select(j);
				Tracer::delay();
				tracer.deselect(j);
				//}
				j--;
			}
			D[i] = D[j];
			//visualizer{
			tracer.patch(i, D[j]);
			Tracer::delay();
			tracer.depatch(i);
			//}
			while (s >= D[i] && i < j) {
				//visualizer{
				tracer.select(i);
				Tracer::delay();
				tracer.deselect(i);
				//}
				i++;
			}
			D[j] = D[i];
			//visualizer{
			tracer.patch(j, D[i]);
			Tracer::delay();
			tracer.depatch(j);
			tracer.deselect(high);
			tracer.deselect(low);
			//}
		}
		D[i] = s;
		//visualizer{
		tracer.patch(i, s);
		Tracer::delay();
		tracer.depatch(i);
		//}
		partition(D, low, i - 1);
		low = i + 1;
	}
}

void quicksort(int D[ARRAYSIZE]) {
	partition(D, 0, ARRAYSIZE - 1);
}

int main()
{
	InitQuickSort();
	//visualizer{
	logger.print("original array = [");
	for (int i = 0; i < ARRAYSIZE-1; i++)
	{
		logger.print(std::to_string(D[i]) + " , ");
	}
	logger.println(std::to_string(D[ARRAYSIZE-1]) + " ] ");
	//}
	quicksort(D);
	//visualizer{
	logger.print("sorted array = [");
	for (int i = 0; i < ARRAYSIZE-1; i++)
	{
		logger.print(std::to_string(D[i]) + " , ");
	}
	logger.println(std::to_string(D[ARRAYSIZE-1]) + " ] ");
	//}
}